### PR TITLE
hashtable index

### PR DESCRIPTION
**Pull request by akiozihao**
#96

### CODE REVIEW

**Comment by yanxiaoqi932** [comment]
May be it's not need to test 2 kinds of index structure in db_test.go, it's enough to test it in bptree_test.go and hashtable_test.go, I think db_test.go is mainly to test db.go.

**Comment by akiozihao** [comment]
> May be it's not need to test 2 kinds of index structure in db_test.go, it's enough to test it in bptree_test.go and hashtable_test.go, I think db_test.go is mainly to test db.go.

ok~

**Comment by yanxiaoqi932** [comment]
OK，I think there art no problem:blush:, maybe we need @roseduan to review the codes further.

File: hashtable.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	options indexOptions
+	tables  []*diskhash.Table
+}
+
+func OpenHashTable(options indexOptions) (*HashTable, error) {
+	tables := make([]*diskhash.Table, options.partitionNum)
+	for i := 0; i < options.partitionNum; i++ {
+		dishHashOptions := diskhash.DefaultOptions
+		dishHashOptions.DirPath = filepath.Join(options.dirPath, fmt.Sprintf(indexFileExt, i))
+		dishHashOptions.SlotValueLength = binary.MaxVarintLen32*3 + binary.MaxVarintLen64
```
Comment (roseduan) [inline hashtable.go:26]: we can explain why the slotVlaueLength is this, to make the code more clear.

File: hashtable.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	g, ctx := errgroup.WithContext(context.Background())
+	for i := range partitionRecords {
+		partition := i
+		if len(partitionRecords[partition]) == 0 {
+			continue
+		}
+		g.Go(func() error {
+			// get the bolt db instance for this partition
```
Comment (roseduan) [inline hashtable.go:59]: bolt db instance -> hashtable instance

File: options.go [hunk 3]
```diff
@@ -48,6 +48,10 @@ type Options struct {
 	// ValueLogFileSize size of a single value log file.
 	// Default value is 1GB.
 	ValueLogFileSize int64
+
+	// indexType.
+	// default value is boltdb.
```
Comment (roseduan) [inline options.go:53]: boltdb -> bptree

File: hashtable.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	for _, table := range ht.tables {
+		err := table.Close()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func encodePosition(cp *wal.ChunkPosition) []byte {
```
Comment (roseduan) [inline hashtable.go:171]: `ChunkPosition` already has a function named Encode

File: hashtable.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+			}
+			return nil
+		})
+	}
+	return g.Wait()
+}
+
+// Sync sync index data to disk
+func (ht *HashTable) Sync() error {
+	return nil
```
Comment (roseduan) [inline hashtable.go:157]: You can create a PR in diskhash to add Sync function

File: hashtable.go (hunk 4 above)
Comment (akiozihao) [reply hashtable.go:171]: `wal.ChunkPostion.Encode()` is not of fixed size. We can use it but need to make slice twice and copy, like 
```
encPos := make([]byte, binary.MaxVarintLen32*3+binary.MaxVarintLen64)
copy(encPos, record.position.Encode())
```

Comment (roseduan) [reply hashtable.go:171]: It`s ok, we can use `bytebufferpool` to avoid allocating every time.

Two similar functions seem a little weird.

**Pull request merged**